### PR TITLE
add memory parity for PL vs Vanilla

### LINKED ISSUE #2080: [test] Add memory parity tests
Preferably add memory parity assertions to existing tests, to avoid extra run time.

Memory parity tests are important to avoid memory leaks.

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

adding memory parity...
Resolves #2080

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: benchmarks/test_basic_parity.py [hunk 1]
```diff
@@ -11,126 +11,164 @@
...
     )
 
+    assert_parity_relative(lightning['memory'], vanilla['memory'], max_diff=max_diff_memory)
 
-def vanilla_loop(cls_model, num_runs=10, num_epochs=10):
+
+def _hook_memory():
+    if torch.cuda.is_available():
+        torch.cuda.synchronize()
+        used_memory = torch.cuda.max_memory_allocated()
```
Comment (Borda) [inline benchmarks/test_basic_parity.py:80]: @SeanNaren, measure...

File: benchmarks/test_basic_parity.py [hunk 2]
```diff
@@ -11,126 +11,164 @@
...
-    for i in tqdm(range(num_runs), desc=f'Vanilla PT with {cls_model.__name__}'):
-        time_start = time.perf_counter()
+    for i in tqdm(range(num_runs), desc=f'{kind} with {cls_model.__name__}'):
+        gc.collect()
+        if device_type == 'cuda':
+            torch.cuda.empty_cache()
+            torch.cuda.reset_max_memory_cached()
+            torch.cuda.reset_max_memory_allocated()
+            torch.cuda.reset_accumulated_memory_stats()
+            torch.cuda.reset_peak_memory_stats()
```
Comment (Borda) [inline benchmarks/test_basic_parity.py:103]: @SeanNaren reset all

File: benchmarks/test_basic_parity.py [hunk 3]
```diff
@@ -11,126 +11,164 @@
...
+    assert np.mean(diffs) < max_diff, f"Lightning {diffs} was worse than vanilla PT (threshold {max_diff})"
+
+
 # ParityModuleMNIST runs with num_workers=1
-@pytest.mark.parametrize('cls_model,max_diff', [
-    (ParityModuleRNN, 0.05),
-    (ParityModuleMNIST, 0.25),  # todo: lower this thr
+@pytest.mark.parametrize('cls_model,max_diff_speed,max_diff_memory', [
+    (ParityModuleRNN, 0.05, 0.0),
+    (ParityModuleMNIST, 0.25, 0.0),  # todo: lower this thr
```
Comment (SeanNaren) [inline benchmarks/test_basic_parity.py:47]: Is it really 0.0 still? I know you investigated a bit was curious. It doesn't seem correct but maybe that's because of how small the memory difference is (memory usage is tiny) Maybe move to a significant figure 1e-5?)

Comment (Borda) [reply benchmarks/test_basic_parity.py:47]: I think so, the model is super small and we run just 4 epochs

File: benchmarks/generate_comparison.py [hunk 4]
```diff
@@ -34,8 +34,8 @@ def _main():
         if os.path.isfile(path_csv):
             df_time = pd.read_csv(path_csv, index_col=0)
         else:
-            vanilla = vanilla_loop(cls_model, num_epochs=NUM_EPOCHS, num_runs=NUM_RUNS)
-            lightning = lightning_loop(cls_model, num_epochs=NUM_EPOCHS, num_runs=NUM_RUNS)
+            vanilla = measure_loops(cls_model, kind="Vanilla PT", num_epochs=NUM_EPOCHS, num_runs=NUM_RUNS)
```
Comment (tchaton) [inline benchmarks/generate_comparison.py:37]: kind="Vanilla PT" -> use_lightning=False

File: benchmarks/test_basic_parity.py (hunk 3 above)
Comment (tchaton) [reply benchmarks/test_basic_parity.py:47]: Not sure adding memory check for such small models make sense.

File: benchmarks/test_basic_parity.py [hunk 5]
```diff
@@ -11,126 +11,164 @@
...
 
-    device = torch.device('cuda' if torch.cuda.is_available() else "cpu")
+    device_type = "cuda" if torch.cuda.is_available() else "cpu"
     torch.backends.cudnn.deterministic = True
-    for i in tqdm(range(num_runs), desc=f'Vanilla PT with {cls_model.__name__}'):
-        time_start = time.perf_counter()
+    for i in tqdm(range(num_runs), desc=f'{kind} with {cls_model.__name__}'):
+        gc.collect()
+        if device_type == 'cuda':
+            torch.cuda.empty_cache()
```
Comment (tchaton) [inline benchmarks/test_basic_parity.py:99]: Can we move this into a function ?

File: benchmarks/test_basic_parity.py (hunk 1 above)
Comment (tchaton) [reply benchmarks/test_basic_parity.py:80]: Won't it be better to make a context manager there ?

File: benchmarks/generate_comparison.py (hunk 4 above)
Comment (Borda) [reply benchmarks/generate_comparison.py:37]: do we want eventually have some more in the future?

File: benchmarks/test_basic_parity.py (hunk 5 above)
Comment (Borda) [reply benchmarks/test_basic_parity.py:99]: this shall be the same for both loops, or why move it in?

File: benchmarks/generate_comparison.py (hunk 4 above)
Comment (tchaton) [reply benchmarks/generate_comparison.py:37]: Don't think so. If we do, we will refactor this part.

**Review by tchaton — APPROVED** [review]

File: benchmarks/generate_comparison.py (hunk 4 above)
Comment (Borda) [reply benchmarks/generate_comparison.py:37]: ok, I'll refactor with the next PR which will add charts for larger models, the generation scrips need refactoring according to this update too...

**Review by SeanNaren — APPROVED** [review]
The memory check for a small model doesn't make sense, but I'd keep it for a future test where we might try a larger model (I think we should).

File: benchmarks/test_basic_parity.py (hunk 3 above)
Comment (awaelchli) [reply benchmarks/test_basic_parity.py:47]: what difference does it make how big the models are?
max_diff_memory there is the difference between the pytorch run and the lightning run with the SAME model. It's perfectly fine if lightning uses the same amount of memory as pytorch. in fact, how would you even explain any other numbers?
There is no logging, no fancy Lightning features, nothing that should occupy extra memory on the gpu. 

also note @Borda is measuring the peak allocated gpu memory since the beginning of the program.  
as an additional sanity check, one could insert an assert that measured_memory > 0

**Review by awaelchli — APPROVED** [review]
LGTM 
see my comment above regarding model size (which does not matter if I understand the tests correctly)
